In [82]:
import pandas as pd
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
import seaborn as sns
import statsmodels.api as sm
from sklearn.preprocessing import PolynomialFeatures
import scipy.stats as stats

df = pd.read_csv('Salarios.csv', decimal = ',')
df_renomeado = df.set_axis(['Anos_estudo', 'Anos_experiencia', 'Idade', 'Salario'], axis=1)
df = df_renomeado
df

Anos_estudo  Anos_experiencia  Idade   Salario
0          18.0               6.0   35.0  16440.58
1          15.0              12.0   39.0  15992.51
2          18.0              11.0   43.0  15009.94
3           2.0               2.0   14.0   6964.13
4           NaN               6.0   18.0   7874.42
5           NaN              16.0   40.0  18329.56
6           NaN               4.0   34.0  15677.25
7           NaN              28.0   42.0   9429.71
8           NaN              27.0   53.0  15743.70
9           NaN              12.0   29.0  15922.81
10          NaN              12.0   37.0  14941.01
11         12.0               NaN   28.0  11938.72
12         11.0               NaN   49.0  15076.47
13          4.0               NaN   34.0  12104.71
14         13.0               NaN   36.0  16353.82
15         11.0               NaN   39.0  15993.23
16          3.0               NaN   16.0   8064.43
17          0.0               NaN   22.0   9667.73
18          9.0               NaN   52.0  18822.42
19          5.0              35.0    NaN  11656.27
20          4.0              32.0    NaN  13155.82
21          4.0              14.0    NaN   9094.41
22          6.0              19.0    NaN  15500.07
23         16.0              16.0    NaN  18964.23
24          5.0               6.0    NaN  14005.99
25         15.0               1.0    NaN  10515.56
26         19.0              29.0   59.0  19837.87
27          5.0               2.0   21.0   8504.59
28          3.0              11.0   27.0  13029.60

In [83]:
df.isna().sum()

Anos_estudo         7
Anos_experiencia    8
Idade               7
Salario             0
dtype: int64

In [84]:
df1 = df
df_filtered = df1.query('Anos_estudo.notnull() and Anos_experiencia.notnull() and Idade.notnull() and Salario.notnull()')
df_filtered

Anos_estudo  Anos_experiencia  Idade   Salario
0          18.0               6.0   35.0  16440.58
1          15.0              12.0   39.0  15992.51
2          18.0              11.0   43.0  15009.94
3           2.0               2.0   14.0   6964.13
26         19.0              29.0   59.0  19837.87
27          5.0               2.0   21.0   8504.59
28          3.0              11.0   27.0  13029.60

In [85]:
df_filtered.isna().sum()

Anos_estudo         0
Anos_experiencia    0
Idade               0
Salario             0
dtype: int64

In [86]:
df_filtered.dtypes

Anos_estudo         float64
Anos_experiencia    float64
Idade               float64
Salario             float64
dtype: object

In [87]:
x = df_filtered[['Anos_estudo', 'Anos_experiencia', 'Idade']]
y = df_filtered['Salario'].to_numpy()
model = LinearRegression()
model.fit(x, y)
score = model.score(x, y)
print(score)

0.9014024420114481


In [88]:
# Vou começar a reconstruir Anos de estudo
# Vou utilizar como variáveis regressoras como Anos de experiência e Idade

df_filtered = df1[ (df1['Anos_estudo'].notnull()) & (df1['Anos_experiencia'].notnull()) & (df1['Idade'].notnull()) & (df1['Salario'].notnull()) ]
x = df_filtered[['Anos_experiencia', 'Idade']].to_numpy()
y = df_filtered['Anos_estudo'].to_numpy()
model_estudo = LinearRegression()
model_estudo.fit(x, y)
score = model_estudo.score(x, y)
score

0.9625591261391816

In [89]:
def completar_estudo(linha):
    if pd.isna(linha['Anos_estudo']) and not pd.isna(linha['Anos_experiencia']) and not pd.isna(linha['Idade']) and not pd.isna(linha['Salario']):
        campos = linha[['Anos_experiencia','Idade']]
        return round(model_estudo.predict([campos])[0])
    return linha['Anos_estudo']

# apply: executar a função chamada 1x para cada linha
df1['Anos_estudo_calculado'] = df1.apply(completar_estudo, axis=1)
df1

Anos_estudo  Anos_experiencia  Idade   Salario  Anos_estudo_calculado
0          18.0               6.0   35.0  16440.58                   18.0
1          15.0              12.0   39.0  15992.51                   15.0
2          18.0              11.0   43.0  15009.94                   18.0
3           2.0               2.0   14.0   6964.13                    2.0
4           NaN               6.0   18.0   7874.42                    0.0
5           NaN              16.0   40.0  18329.56                   12.0
6           NaN               4.0   34.0  15677.25                   17.0
7           NaN              28.0   42.0   9429.71                    3.0
8           NaN              27.0   53.0  15743.70                   15.0
9           NaN              12.0   29.0  15922.81                    5.0
10          NaN              12.0   37.0  14941.01                   13.0
11         12.0               NaN   28.0  11938.72                   12.0
12         11.0               NaN   49.0  15076.47                   11.0
13          4.0               NaN   34.0  12104.71                    4.0
14         13.0               NaN   36.0  16353.82                   13.0
15         11.0               NaN   39.0  15993.23                   11.0
16          3.0               NaN   16.0   8064.43                    3.0
17          0.0               NaN   22.0   9667.73                    0.0
18          9.0               NaN   52.0  18822.42                    9.0
19          5.0              35.0    NaN  11656.27                    5.0
20          4.0              32.0    NaN  13155.82                    4.0
21          4.0              14.0    NaN   9094.41                    4.0
22          6.0              19.0    NaN  15500.07                    6.0
23         16.0              16.0    NaN  18964.23                   16.0
24          5.0               6.0    NaN  14005.99                    5.0
25         15.0               1.0    NaN  10515.56                   15.0
26         19.0              29.0   59.0  19837.87                   19.0
27          5.0               2.0   21.0   8504.59                    5.0
28          3.0              11.0   27.0  13029.60                    3.0

In [92]:
# Vou começar a reconstruir Anos de experiência
# Vou utilizar como variáveis regressoras como Anos de estudo calculado e Idade

df_filtered = df1[ (df1['Anos_experiencia'].notnull()) & (df1['Idade'].notnull()) & (df1['Salario'].notnull()) & (df1['Anos_estudo_calculado'].notnull()) ]
x = df_filtered[['Idade', 'Anos_estudo_calculado']].to_numpy()
y = df_filtered['Anos_experiencia'].to_numpy()
model_experiencia = LinearRegression()
model_experiencia.fit(x, y)
score = model_experiencia.score(x, y)
score

0.9852579396677799

In [93]:
def completar_experiencia(linha):
    if pd.isna(linha['Anos_experiencia']) and not pd.isna(linha['Idade']) and not pd.isna(linha['Salario']) and not pd.isna(linha['Anos_estudo_calculado']):
        campos = linha[['Idade', 'Anos_estudo_calculado']]
        return round(model_experiencia.predict([campos])[0])
    return linha['Anos_experiencia']

# apply: executar a função chamada 1x para cada linha
df1['Anos_experiencia_calculado'] = df1.apply(completar_experiencia, axis=1)
df1

Anos_estudo  Anos_experiencia  Idade   Salario  Anos_estudo_calculado  \
0          18.0               6.0   35.0  16440.58                   18.0   
1          15.0              12.0   39.0  15992.51                   15.0   
2          18.0              11.0   43.0  15009.94                   18.0   
3           2.0               2.0   14.0   6964.13                    2.0   
4           NaN               6.0   18.0   7874.42                    0.0   
5           NaN              16.0   40.0  18329.56                   12.0   
6           NaN               4.0   34.0  15677.25                   17.0   
7           NaN              28.0   42.0   9429.71                    3.0   
8           NaN              27.0   53.0  15743.70                   15.0   
9           NaN              12.0   29.0  15922.81                    5.0   
10          NaN              12.0   37.0  14941.01                   13.0   
11         12.0               NaN   28.0  11938.72                   12.0   
12         11.0               NaN   49.0  15076.47                   11.0   
13          4.0               NaN   34.0  12104.71                    4.0   
14         13.0               NaN   36.0  16353.82                   13.0   
15         11.0               NaN   39.0  15993.23                   11.0   
16          3.0               NaN   16.0   8064.43                    3.0   
17          0.0               NaN   22.0   9667.73                    0.0   
18          9.0               NaN   52.0  18822.42                    9.0   
19          5.0              35.0    NaN  11656.27                    5.0   
20          4.0              32.0    NaN  13155.82                    4.0   
21          4.0              14.0    NaN   9094.41                    4.0   
22          6.0              19.0    NaN  15500.07                    6.0   
23         16.0              16.0    NaN  18964.23                   16.0   
24          5.0               6.0    NaN  14005.99                    5.0   
25         15.0               1.0    NaN  10515.56                   15.0   
26         19.0              29.0   59.0  19837.87                   19.0   
27          5.0               2.0   21.0   8504.59                    5.0   
28          3.0              11.0   27.0  13029.60                    3.0   

    Anos_experiencia_calculado  
0                          6.0  
1                         12.0  
2                         11.0  
3                          2.0  
4                          6.0  
5                         16.0  
6                          4.0  
7                         28.0  
8                         27.0  
9                         12.0  
10                        12.0  
11                         4.0  
12                        27.0  
13                        18.0  
14                        11.0  
15                        16.0  
16                         1.0  
17                        10.0  
18                        32.0  
19                        35.0  
20                        32.0  
21                        14.0  
22                        19.0  
23                        16.0  
24                         6.0  
25                         1.0  
26                        29.0  
27                         2.0  
28                        11.0

In [95]:
df_filtered = df1[ (df1['Idade'].notnull()) & (df1['Salario'].notnull()) & (df1['Anos_estudo_calculado'].notnull()) & (df1['Anos_experiencia_calculado'].notnull()) ]
x = df_filtered[['Anos_estudo_calculado', 'Anos_experiencia_calculado']].to_numpy()
y = df_filtered['Idade'].to_numpy()
model_idade = LinearRegression()
model_idade.fit(x, y)
score = model_idade.score(x, y)
score

0.9950410158423627

In [96]:
def completar_idade(linha):
    if pd.isna(linha['Idade']) and not pd.isna(linha['Salario']) and not pd.isna(linha['Anos_estudo_calculado']) and not pd.isna(linha['Anos_experiencia_calculado']):
        campos = linha[['Anos_estudo_calculado', 'Anos_experiencia_calculado']]
        return round(model_idade.predict([campos])[0])
    return linha['Idade']

# apply: executar a função chamada 1x para cada linha
df1['Idade_calculado'] = df1.apply(completar_idade, axis=1)
df1

Anos_estudo  Anos_experiencia  Idade   Salario  Anos_estudo_calculado  \
0          18.0               6.0   35.0  16440.58                   18.0   
1          15.0              12.0   39.0  15992.51                   15.0   
2          18.0              11.0   43.0  15009.94                   18.0   
3           2.0               2.0   14.0   6964.13                    2.0   
4           NaN               6.0   18.0   7874.42                    0.0   
5           NaN              16.0   40.0  18329.56                   12.0   
6           NaN               4.0   34.0  15677.25                   17.0   
7           NaN              28.0   42.0   9429.71                    3.0   
8           NaN              27.0   53.0  15743.70                   15.0   
9           NaN              12.0   29.0  15922.81                    5.0   
10          NaN              12.0   37.0  14941.01                   13.0   
11         12.0               NaN   28.0  11938.72                   12.0   
12         11.0               NaN   49.0  15076.47                   11.0   
13          4.0               NaN   34.0  12104.71                    4.0   
14         13.0               NaN   36.0  16353.82                   13.0   
15         11.0               NaN   39.0  15993.23                   11.0   
16          3.0               NaN   16.0   8064.43                    3.0   
17          0.0               NaN   22.0   9667.73                    0.0   
18          9.0               NaN   52.0  18822.42                    9.0   
19          5.0              35.0    NaN  11656.27                    5.0   
20          4.0              32.0    NaN  13155.82                    4.0   
21          4.0              14.0    NaN   9094.41                    4.0   
22          6.0              19.0    NaN  15500.07                    6.0   
23         16.0              16.0    NaN  18964.23                   16.0   
24          5.0               6.0    NaN  14005.99                    5.0   
25         15.0               1.0    NaN  10515.56                   15.0   
26         19.0              29.0   59.0  19837.87                   19.0   
27          5.0               2.0   21.0   8504.59                    5.0   
28          3.0              11.0   27.0  13029.60                    3.0   

    Anos_experiencia_calculado  Idade_calculado  
0                          6.0             35.0  
1                         12.0             39.0  
2                         11.0             43.0  
3                          2.0             14.0  
4                          6.0             18.0  
5                         16.0             40.0  
6                          4.0             34.0  
7                         28.0             42.0  
8                         27.0             53.0  
9                         12.0             29.0  
10                        12.0             37.0  
11                         4.0             28.0  
12                        27.0             49.0  
13                        18.0             34.0  
14                        11.0             36.0  
15                        16.0             39.0  
16                         1.0             16.0  
17                        10.0             22.0  
18                        32.0             52.0  
19                        35.0             51.0  
20                        32.0             47.0  
21                        14.0             30.0  
22                        19.0             37.0  
23                        16.0             44.0  
24                         6.0             23.0  
25                         1.0             29.0  
26                        29.0             59.0  
27                         2.0             21.0  
28                        11.0             27.0

In [97]:
# Recalculo a minha regressão original utilizando as variáveis novas
x = df1[['Anos_estudo_calculado', 'Anos_experiencia_calculado', 'Idade_calculado']]
y = df1['Salario'].to_numpy()
model = LinearRegression()
model.fit(x, y)
score = model.score(x, y)
print(score)

0.6485333536793012


* A regressão Linear sem o tratamento dos missings, obteve um r^2 de 0.90
* Já para regressão com os missings tratados, obteve uma significância menor com um r^2 de 0.64, mas sem viés e um poder de explicação maior
* A diferença percentual no poder de explicação entre os dois coeficientes de determinação é aproximadamente 28.88%
* Sugerindo que o modelo ajustado com os dados faltantes tratados é mais robusto e generalizável, resultando em um modelo mais representativo